src/cachingreaderchunk.h

```diff
@@ -0,0 +1,141 @@
+// Each chunk holds a fixed number kFrames of frames with samples for
+// kChannels.
+// The class is not thread-safe although it is shared between CachingReader
+// and CachingReaderWorker! A lock-free FIFO ensures that only a single
+// thread has exclusive access on each chunk. This abstract base class
+// is available for both the worker thread and the cache.
+class CachingReaderChunkWorker {
+public:
+    static const SINT kDefaultIndex;
+    static const SINT kChannels;
+    static const SINT kFrames;
+    static const SINT kSamples;
+
+    // Returns the corresponding chunk index for a frame index
+    inline static SINT indexForFrame(SINT frameIndex) {
+        DEBUG_ASSERT(Mixxx::AudioSource::getMinFrameIndex() <= frameIndex);
+        const SINT chunkIndex = frameIndex / kFrames;
+        return chunkIndex;
+    }
+
+    // Returns the corresponding chunk index for a frame index
+    inline static SINT frameForIndex(SINT chunkIndex) {
+        DEBUG_ASSERT(0 <= chunkIndex);
+        return chunkIndex * kFrames;
+    }
+
+    // Converts frames to samples
+    inline static SINT frames2samples(SINT frames) {
+        return frames * kChannels;
+    }
+    // Converts samples to frames
+    inline static SINT samples2frames(SINT samples) {
+        DEBUG_ASSERT(0 == (samples % kChannels));
+        return samples / kChannels;
+    }
+
+    virtual ~CachingReaderChunkWorker();
+
+    SINT getIndex() const {
+        return m_index;
+    }
+
+    bool isValid() const {
+        return 0 <= getIndex();
+    }
+
+    SINT getFrameCount() const {
+        return m_frameCount;
+    }
+
+    // Check if the audio source has sample data available
+    // for this chunk.
+    bool isReadable(
+            const Mixxx::AudioSourcePointer& pAudioSource,
+            SINT maxReadableFrameIndex) const;
+
+    // Read sample frames from the audio source and return the
+    // number of frames that have been read. The in/out parameter
+    // pMaxReadableFrameIndex is adjusted if reading fails.
+    SINT readSampleFrames(
+            const Mixxx::AudioSourcePointer& pAudioSource,
+            SINT* pMaxReadableFrameIndex);
+
+    // Copy sampleCount samples starting at sampleOffset from
+    // the chunk's internal buffer into sampleBuffer.
+    void copySamples(
+            CSAMPLE* sampleBuffer,
+            SINT sampleOffset,
+            SINT sampleCount) const;
+
+protected:
+    explicit CachingReaderChunkWorker(CSAMPLE* sampleBuffer);
+
+    void init(SINT index);
+
+private:
+    volatile SINT m_index;
+
+    // The worker thread will fill the sample buffer and
+    // set the frame count.
+    CSAMPLE* const m_sampleBuffer;
+    volatile SINT m_frameCount;
+};
+
+// The derived class is only accessible for the cache, but not the
+// worker thread. The state READ_PENDING indicates that the worker
+// thread is in control.
+class CachingReaderChunk: public CachingReaderChunkWorker {
```

Comment:
The current state is a bit confusing, because first I thought CachingReaderChunkWorker is a thread. 
I think I got why you split up the class: to separate the function that are accessed from different threads. 
So it is probably only a naming issue. 

C++ does not enforce this restrictions yet, right?  Maybe we can manage it by accessing the chunks via two different
abstract classes. Like ICachinReaderChunkForWorker  and ICachinReaderChunkForCache